ballerina/variable_validator.bal

```diff
@@ -154,4 +172,47 @@ class VariableValidator {
             self.errors.push(getErrorDetailRecord(message, location)); 
         }
     }
+
+    isolated function checkVariableUsageCompatibility(__InputValue[] input, parser:VariableDefinition varDef,
+                                                      parser:ArgumentNode argNode) {
+        __InputValue? inputValue = getInputValueFromArray(input, argNode.getName());
+        __Type? actualType = self.getTypeRecordFromTypeName(varDef.kind);
+        if inputValue is __InputValue && actualType is __Type {
+            if getOfType(inputValue.'type)?.name != getOfType(actualType)?.name {
+                string message = string`Variable "${<string>argNode.getVariableName()}" of type "${varDef.kind}" `+
+                string`used in position expecting type "${getTypeNameFromType(inputValue.'type)}".`;
+                self.errors.push(getErrorDetailRecord(message, argNode.getLocation()));
+            } else if inputValue.'type.kind == NON_NULL && actualType.kind != NON_NULL && varDef?.defaultValue is () {
+                string message = string`Variable "${<string>argNode.getVariableName()}" of type "${varDef.kind}" `+
+                string`used in position expecting type "${getTypeNameFromType(inputValue.'type)}".`;
+                self.errors.push(getErrorDetailRecord(message, argNode.getLocation()));
+            }
+        }
+    }
+
+    isolated function getTypeRecordFromTypeName(string typeName) returns __Type? {
+        //TODO: Improve for 'List' type variables
+        if typeName.endsWith("!") {
+            __Type wrapperType = {
+                kind: NON_NULL
+            };
+            string ofTypeName = getOfTypeNameFromTypeName(typeName);
+            __Type? ofType = getTypeFromTypeArray(self.schema.types, ofTypeName);
+            if ofType is __Type {
+                wrapperType.ofType = ofType;
+            }
+            return wrapperType;
+        } else {
+            return self.getSchemaTypeFromTypeArray(self.schema.types, typeName);
+        }
+    }
+
+    isolated function getSchemaTypeFromTypeArray(__Type[] types, string typeName) returns __Type? {
```

Comment:
We already have `getTypeFromTypeArray` function. Why do we need to duplicate it?